tasks/eth/021-holocene-protocol-versions/SignFromJson.s.sol

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.15;
+
+import {SignFromJson as OriginalSignFromJson} from "script/SignFromJson.s.sol";
+import {Simulation} from "@base-contracts/script/universal/Simulation.sol";
+import {Proxy} from "@eth-optimism-bedrock/src/universal/Proxy.sol";
+import {SystemConfig} from "@eth-optimism-bedrock/src/L1/SystemConfig.sol";
+import {ProtocolVersions, ProtocolVersion} from "@eth-optimism-bedrock/src/L1/ProtocolVersions.sol";
+import {Types} from "@eth-optimism-bedrock/scripts/Types.sol";
+import {console2 as console} from "forge-std/console2.sol";
+import {stdToml} from "forge-std/StdToml.sol";
+import {Vm, VmSafe} from "forge-std/Vm.sol";
+import {LibString} from "solady/utils/LibString.sol";
+
+contract SignFromJson is OriginalSignFromJson {
+    using LibString for string;
+
+    // Chains for this task.
+    string constant l1ChainName = "mainnet";
+    string constant l2ChainName = "op";
+
+    uint256 constant protoVerHolocene = 0x0000000000000000000000000000000000000009000000000000000000000000;
+
+    // Safe contract for this task.
+    address foundationUpgradesSafe = vm.envAddress("FOUNDATION_SAFE");
+
+    // All L1 proxy addresses.
+    Types.ContractSet proxies;
+
+    function setUp() public {
+        proxies = _getContractSet();
+    }
+
+    function getAllowedStorageAccess() internal view override returns (address[] memory allowed) {
+        allowed = new address[](2);
+        allowed[0] = foundationUpgradesSafe;
+        allowed[1] = proxies.ProtocolVersions;
+    }
+
+    /// @notice Checks the correctness of the deployment
+    function _postCheck(Vm.AccountAccess[] memory accesses, Simulation.Payload memory /* simPayload */ )
+        internal
+        view
+        override
+    {
+        console.log("Running assertions");
+
+        checkStateDiff(accesses);
+        checkProtocolVersions();
+
+        console.log("All assertions passed!");
+    }
+
+    function getCodeExceptions() internal view override returns (address[] memory exceptions) {
+        // No exceptions are expected in this task, but it must be implemented.
+    }
+
+    function checkProtocolVersions() internal view {
+        console.log("Checking ProtocolVersions at ", proxies.ProtocolVersions);
+        ProtocolVersions pv = ProtocolVersions(proxies.ProtocolVersions);
+        require(pv.owner() == foundationUpgradesSafe, "PV owner must be Foundation Upgrade Safe");
+        // We ignore the recommended version, as it is set by this task
+        require(ProtocolVersion.unwrap(pv.required()) == protoVerHolocene, "Required PV must be Holocene");
+        require(ProtocolVersion.unwrap(pv.recommended()) == protoVerHolocene, "Required PV must be Holocene");
+    }
+
+    /// @notice Reads the contract addresses from the superchain registry.
+    function _getContractSet() internal view returns (Types.ContractSet memory _proxies) {
```

Comment:
If you use the `SuperchainRegistry` module, you'd have the proxies available at `proxies`.

I realize however that you wouldn't need the standard versions. Maybe you could split the current `SuperchainRegistry` contract into two, like `SuperchainRegistryConfig` and `SuperchainRegistryStandardVersions`, and then here you would only use the `SuperchainRegistryConfig`.

You could then bundle them in the current contract like
```solidity
contract SuperchainRegistry is SuperchainRegistryConfig, SuperchainRegistryStandardVersions {
    constructor(string memory _l1ChainName, string memory _l2ChainName, string memory _opContractsRelease)
    SuperchainRegistryConfig(_l1ChainName, _l2ChainName)
    SuperchainRegistryStandardVersions(_l1ChainName, _opContractsRelease)
    {}
}
``

Reply:
I like this idea but prefer to kick it into future work in the interests of time.